arrow-syntax/src/main/kotlin/arrow/syntax/function/memoization.kt

```diff
@@ -214,6 +215,24 @@ private data class MemoizeKey22<out P1, out P2, out P3, out P4, out P5, out P6,
 }
 
 private class MemoizedHandler<F, in K : MemoizedCall<F, R>, out R>(val f: F) {
-  private val m = newConcurrentMap<K, R>()
-  operator fun invoke(k: K): R = m[k] ?: run { m.putSafely(k, k(f)) }
+  private val cache = atomic(emptyMap<K, R>())
+  operator fun invoke(k: K): R {
+    val cached = cache.value[k]
+    // No cached value found, compute one
+    return if (cached == null) {
+      val b = k(f)
+      cache.loop { old ->
+        val bb = old[k]
+        // No value is set yet (race condition)
+        if (bb == null) {
+          // Value was set, return the value
+          if (cache.compareAndSet(old, old + Pair(k, b))) return b
+          // keep looping failed to set
+          else Unit
+          // A value was already set first, return it
```

Comment:
We should add some concurrent tests to check for this branching.

Reply:
Changes done in https://github.com/arrow-kt/arrow-core/pull/331/commits/0dbc7145665524c206f7daf7b3dfead30344dbea 

I added some docs and this single test

Reply:
👌